src/infer_request.cc

```diff
@@ -142,49 +144,55 @@ InferenceRequest::SetState(InferenceRequest::State new_state)
     std::stringstream ss;
     ss << LogRequest() << "Invalid request state transition from " << state_
        << " to " << new_state;
-    return Status(Status::Code::INVALID_ARG, ss.str());
+    return Status(Status::Code::INTERNAL, ss.str());
   };
 
   // Define state transitions
   switch (state_) {
     case InferenceRequest::State::INITIALIZED: {
-      if (new_state != InferenceRequest::State::STARTED) {
+      if (new_state != InferenceRequest::State::PENDING) {
         return generate_error();
       }
-      state_ = new_state;
       IncrementPendingRequestCount();
       break;
     }
-    case InferenceRequest::State::STARTED: {
+    case InferenceRequest::State::PENDING: {
       if (new_state != InferenceRequest::State::EXECUTING) {
         return generate_error();
       }
-      state_ = new_state;
       DecrementPendingRequestCount();
       break;
     }
     case InferenceRequest::State::EXECUTING: {
       if (new_state != InferenceRequest::State::RELEASED) {
         return generate_error();
       }
-      state_ = new_state;
       break;
     }
     case InferenceRequest::State::RELEASED: {
-      // No state transition currently supported after release.
-      return generate_error();
+      if (new_state != InferenceRequest::State::INITIALIZED) {
+        // Only transition currently supported after release is to start over
+        // again, such as re-using request objects for multiple inferences.
+        return generate_error();
+      }
```

Comment:
As the time between release and destruction is not specified - I think we would want to decrement pending count on release and not in the destructor.

So overall, the logic would be:

1) Initialize -> set is_pending to false
2) Initialize -> Pending -> set is_pending to true and Increment
3) Pending -> Execute -> set is_pending to false and decrement
4) *-> Released -> if (is_pending) set is_pending to false and decrement 

it is also possible I believe that it can all be handled in the state method without the need for the pending flag - if we check on which state is before release (pending -> release decrement, otherwise do not)

1. Initialized -> Pending: Increment
2. Pending -> Execute: Decrement
3. Pending -> Released: Decrement
4. (other) -> Released: Do nothing

I'd like to see if we can use the state or the pending flag - and not both if possible.

Reply:
> I'd like to see if we can use the state or the pending flag - and not both if possible.

I originally had decrement in destructor as a RAII-style sanity check that we'll clean up no matter what code path may get hit. However, I think your sentiment makes sense that we can capture the intended code paths in the transitions, such as Pending->Released for early exits. I'll try this approach within states rather than having an additional flag.

Reply:
@nnshah1 moved logic into state transitions only, appears to be working and passing CI as well. Let me know what you think.

Reply:
LGTM -